Source/cryptalgo/SecureSocketPort.cpp

```diff
@@ -230,34 +262,44 @@ void SecureSocketPort::Handler::Update() {
 
         ASSERT(_ssl != nullptr);
 
+        // Client
         if (_handShaking == CONNECTING) {
             if ((result = SSL_connect(static_cast<SSL*>(_ssl))) == 1) {
-                _handShaking = EXCHANGE;
+                _handShaking = CONNECTED;
+                // If server has sent a certificate, and, we want to do 'our' own check
+                ValidateHandShake();
             }
-        }
+        } // Server
         else if (_handShaking == ACCEPTING) {
             if ((result = SSL_accept(static_cast<SSL*>(_ssl))) == 1) {
-                _handShaking = EXCHANGE;
+                _handShaking = CONNECTED;
+                // Client has sent a certificate (optional, on request only)
+                ValidateHandShake();
             }
-        }
-        
+        } // Re-initialie a previous session
         if (_handShaking == EXCHANGE) {
             if ((result = SSL_do_handshake(static_cast<SSL*>(_ssl))) == 1) {
                 ValidateHandShake();
+                _handShaking = CONNECTED;
             }
         }
 
         if (result != 1) {
             result = SSL_get_error(static_cast<SSL*>(_ssl), result);
-            if ((result != SSL_ERROR_WANT_READ) && (result != SSL_ERROR_WANT_WRITE)) {
-                _handShaking = ERROR;
-            }
-            else if (result == SSL_ERROR_WANT_WRITE) {
+
+            if ((result == SSL_ERROR_WANT_READ) || (result == SSL_ERROR_WANT_WRITE)) {
+                // Non-blocking I/O: select may be used to check if condition has changed
```

Comment:
If SSL still needs another READ or WRITE to complete the Handshaking, I assume t has *no* use to ValidateHandShake () and just wait for more data and "stay" in the EXCHANGE state...
I can not imagine that if the Error of SSL is, I need more data to write or read that the handshaking is over and thus that you can validate the handshake and even worse that you can move the connection to CONNECTED as I assume the exchange for TLS is still in progress..